### ISSUE TITLE
argument minRows in setGradientBoosting machine described wrongly and does not allow the recommended default value used by xgboost

### ISSUE THREAD

**Issue by egillax** [OHDSI/PatientLevelPrediction#256]
**Describe the bug**
the value minRows has a default value of 2 and is described thus in the setGradientBoostingMachine function:

> @param minRows   The minimum number of rows required at each end node of the tree

minRows seems to translate to the parameter min_child_weight in xgboost (see [here](https://xgboost.readthedocs.io/en/stable/parameter.html))

Which is described as:

> Minimum sum of instance weight (hessian) needed in a child. 

This is not the minimum number of rows required at the leafs except in the case of linear regression with squared loss, which we don't use in PLP. So I think we should fix the documentation, and rename the variable to minChildWeight so it reflects better what it is.

Further there is a check in the code on the minRows variable:

```r
if(!class(minRows) %in% c("numeric", "integer"))
    stop('minRows must be a numeric value >1')
if(sum(minRows < 2)>0)
    stop('minRows must be greater that 1')
```
The recommended value by xgboost for min_child_weight is 1, which according to this is not allowed in PLP. The allowed range of it is from [0, infinity). My suggestions are that we should have the default value for this value the same as in xgboost, 1. And only check if this value is bigger than zero.

I can implement those changes if you agree @jreps.

Cheers, Egill

**Comment by jreps** [comment]
sounds good.  Feel free to add any other hyper-parameters as well (I think there are many we don't let the user specify right now, but if you see any you think are worth letting users tune please add them).

**Comment by egillax** [comment]
I did a little bit of testing of additional hyperparameters. I ran about 5400 model fits on a cohort I have (300k target, 1.8% outcome rate, default features). Then I fit a random forest regression model to predict the AUC using the hyperparameters as my predictors. Then I can calculate feature importance with permutations to see which hyperparameters are affecting the fit the most. The results are:

- ntrees, maxDepth and learnRate are by far the most important with about 23-26% drop in mean squared error when they are permuted.
- minChildWeight (previously minRows) has about 6.8% drop
- lambda, which is l2 regularization has about 4.5% drop.
- alpha (l1 regularization) has 0.7% drop.
- others had less effect

I'd suggest at least having those. So lambda and alpha would be new. Maybe the lower three (minChildWeight, lambda and alpha) could just have the default value as a default and the top three ones can be used in the default gridsearch. 

Currently the grid search involves 12 parameter combinations:

ntrees(100, 1000) , learnRate(0.005, 0.01, 0.1) and maxDepth(4, 6, 17) 

I do feel some of those values might not be optimal. I think 1000 trees is a bit of a jump from 100 trees which is the default, I would suggest something lower like 200 or 300, except if you have evidence that 1000 trees are sometimes chosen?  The default xgboost value for learnRate is 0.3, which I feel should be included in the gridsearch at the expense of removing one or even both of the lower ones. I also feel 17 maxDepth is quite a jump from the default of 6. (The optimal hyperparameter for my test were ntrees=200, maxDepth=5,learnRate=0.1, lambda=5 and alpha=3)

So my suggestions for the hyperparameter search would be (overall trying to stray more conservatively from the defaults):

ntrees(100, 300), learnRate(0.05, 0.1, 0.3) and maxDepth(4, 6, 8)

If there are results somewhere where I could dig in a bit about what hyperparameters have been chosen for different datasets I'd love to see those.

Then there is one parameter called scale_pos_weight to put more weight on the positive class in the loss function. I've seen people commonly use it but in my test it caused worse performance. I'd still suggest having it as an option with default value of 1 and then removing the row-wise weights that are currently used but cause a warning that they are possibly not being used, I think there is an issue in the xgboost R documentation which says that the weight argument should be used as a parameter in xgb.train but according to [this](https://github.com/dmlc/xgboost/issues/7587) it should be used in xgb.Dmatrix. To me it sounds simpler to have the weigh the positive class in the loss rather than weighting by row. 

I could submit a pull request for these changes tomorrow.

**Comment by jreps** [comment]
awesome!  Yeah - please make these changes (modifying the current default values and adding extra hyperparameters that seem to make an impact).

Also - happy for you to use scale_pos_weight instead of the current weights (that warning is annoying and I will be glad for it to be removed :) ).
